ports/gstplugin/servosrc.rs

```diff
@@ -0,0 +1,566 @@
+            let origin = Point2D::origin();
+            debug_assert_eq!(gfx.gl.get_error(), gl::NO_ERROR);
+            EmbedderCoordinates {
+                hidpi_factor: Scale::new(1.0),
+                screen: size,
+                screen_avail: size,
+                window: (size, origin),
+                framebuffer: size,
+                viewport: Rect::new(origin, size),
+            }
+        })
+    }
+
+    fn set_animation_state(&self, _: AnimationState) {}
+
+    fn get_gl_context(&self) -> servo_media::player::context::GlContext {
+        servo_media::player::context::GlContext::Unknown
+    }
+
+    fn get_native_display(&self) -> servo_media::player::context::NativeDisplay {
+        servo_media::player::context::NativeDisplay::Unknown
+    }
+
+    fn get_gl_api(&self) -> servo_media::player::context::GlApi {
+        servo_media::player::context::GlApi::OpenGL3
+    }
+}
+
+impl ObjectSubclass for ServoSrc {
+    const NAME: &'static str = "ServoSrc";
+    // gstreamer-gl doesn't have support for GLBaseSrc yet
+    // https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/issues/219
+    type ParentType = BaseSrc;
+    type Instance = ElementInstanceStruct<Self>;
+    type Class = ClassStruct<Self>;
+
+    fn new() -> Self {
+        let (sender, receiver) = crossbeam_channel::bounded(1);
+        thread::spawn(move || ServoThread::new(receiver).run());
+        let (acks, ackr) = crossbeam_channel::bounded(1);
+        let _ = sender.send(ServoSrcMsg::GetSwapChain(acks));
+        let swap_chain = ackr.recv().expect("Failed to get swap chain");
+        let info = Mutex::new(None);
+        Self {
+            sender,
+            swap_chain,
+            info,
+        }
+    }
+
+    fn class_init(klass: &mut ClassStruct<Self>) {
+        klass.set_metadata(
+            "Servo as a gstreamer src",
+            "Filter/Effect/Converter/Video",
+            "The Servo web browser",
+            env!("CARGO_PKG_AUTHORS"),
+        );
+
+        let src_caps = Caps::new_simple(
+            "video/x-raw",
+            &[
+                ("format", &VideoFormat::Bgrx.to_string()),
+                ("width", &IntRange::<i32>::new(512, 1024)),
+                ("height", &IntRange::<i32>::new(512, 1024)),
+                (
+                    "framerate",
+                    &FractionRange::new(Fraction::new(1, 1), Fraction::new(15, 1)),
+                ),
+            ],
+        );
+        let src_pad_template =
+            PadTemplate::new("src", PadDirection::Src, PadPresence::Always, &src_caps).unwrap();
+        klass.add_pad_template(src_pad_template);
+    }
+
+    glib_object_subclass!();
+}
+
+impl ObjectImpl for ServoSrc {
+    glib_object_impl!();
+
+    fn constructed(&self, obj: &glib::Object) {
+        self.parent_constructed(obj);
+        let basesrc = obj.downcast_ref::<BaseSrc>().unwrap();
+        basesrc.set_live(true);
+        basesrc.set_format(Format::Time);
```

Comment:
You probably also want `basesrc.do_timestamp(true)` here until you implement more accurate timestamping yourself

Reply:
```
   --> ports/gstplugin/servosrc.rs:454:10
    |
454 |     basesrc.do_timestamp(true);
    |             ^^^^^^^^^^^^ method not found in `&gstreamer_base::auto::base_src::BaseSrc`
```

Reply:
Sorry, `set_do_timestamp(true)`, not `do_timestamp(true)` :)